docs: add a glossary (#586)
pejulian/romcal: docs/glossary.md

# Glossary

This glossary provides definitions of terms used in the context of this romcal project. The terms are listed in alphabetical order.

> 

Reviewer: Guys, feel free to modify the suggestion as you see fit.

```suggestion

The structure of a glossary items is as follows:

- the term name in a heading of the second level, without any abbreviation:
   - abbreviations, acronyms, initialisms (etc) should be made a separate item with `See [Full Term Name](#full-term-name).` (change the link name and anchor as required);
   - the headings should be ordered alphabetically;
- definition of the term:
   - ideally, it should full in a single, short paragraph;
   - provide an abbreviation (if it is used) after the first mention of the term;
- add a list named `See also:` where the synonyms and related items would be provided (provide as many links as required; prefer links to the glossary items). 

```

Reviewer 2: i don't think we need to call out "blank lines". as that's a formatting requirement for proper line returns that are part of markdown that we'd have to look out for. they aren't necessary for our format.

Reviewer: Okay, I have remove the list item talking about blank lines from my suggestion.

## Celebration

A celebration is a liturgical or sacramental event, typically consisting of rituals, readings, prayers, and often the Eucharist (Mass), whether or not it is associated with a liturgical day.

> [!NOTE]

> > Currently, Romcal does not provide specific information about celebrations, such as the list of Masses for a given day or the hours of the Divine Office. Instead, it indicates unique liturgical days determined within a liturgical calendar. However, the ability to list these celebrations—linked to a liturgical day or corresponding to a civil day (such as Vigil Masses)—is a feature that may be added in the future.

Reviewer: While I prefer calling the LotH as the _Divine Office_, I think it is usually called so in the UK (and the Former British Colonies). As we use American English as the base of the romcal l10n, I suggest to prefer to use the `Liturgy of the Hours` name instead.

```suggestion
> Currently, romcal does not provide specific information about celebrations, such as the list of Masses for a given day or the hours of the Liturgy of the Hours. Instead, it indicates unique liturgical days determined within a liturgical calendar. However, the ability to list these celebrations—linked to a liturgical day or corresponding to a civil day (such as Vigil Masses)—is a feature that may be added in the future.
```

Reviewer 2: both Divine Office & Liturgy of the Hours are [accepted names](https://www.usccb.org/prayer-and-worship/liturgy-of-the-hours)

Reviewer 2: if it was just the romcal name change, i'm down for changing all the casing to lower case. I don't think we need to change this everywhere to match American English, given that we don't have a standard way of referencing LotH vs DO, and we should just work to have what is needed, and add more in successive PRs.

it's a good thing to have a bunch of iterations on adding and tweaking documentation. it doesn't have to be perfect the first commit that goes in. this is the change to add, not perfect, a glossary.

Author: Totally agree!

Reviewer 2: i've fixed the romcal name change, and i think 2 of us agree we want to leave Divine Office as is, and we'll update referencing in the future on this if it's agreed as a group.

Author: For me, this works (and the name is shorter). And thank you for moving the work forward in this PR 🙏

> 

Reviewer: ```suggestion

## Divine Office

See the [Liturgy of the Hours](#liturgy-of-the-hours).

```

An English translation of the GILH can be found [here](https://www.liturgyoffice.org.uk/Resources/Rites/GILH.pdf).

## General Instruction on the Roman Missal (GIRM)

The General Instruction on the Roman Missal (GIRM) is the detailed document that provides the rubrics for the celebration of the Mass. It is an official liturgical document that was promulgated by the Congregation for Divine Worship and the Discipline of the Sacraments.

An English translation of the GIRM can be found [here](https://www.liturgyoffice.org.uk/Resources/GIRM/Documents/GIRM.pdf).

## General Norms for the Liturgical Year and the Calendar (GNLYC)

The General Norms for the Liturgical Year and the Calendar (GNLYC) is the document that provides the general norms for the liturgical year and the calendar. It is an official liturgical document that was promulgated by the Congregation for Divine Worship and the Discipline of the Sacraments.

An English translation of the GNLYC can be found [here](https://www.liturgyoffice.org.uk/Resources/GIRM/Documents/GNLY.pdf).

> 

Reviewer: I think we should only mention an older title of the document. However, I don’t think it is so important that we should add a separate glossary item for this pointing back to this item.

```suggestion

Older English name of the document was [_Universal Norms on the Liturgical Year and General Roman Calendar_](https://www.liturgyoffice.org.uk/Calendar/Info/GNLY.pdf), sometimes abbreviated as UNLY, UNLYC, UNLYGRC.

```

## Liturgical Color

The liturgical colors are the colors used in the liturgical vestments and altar cloths for the celebration of the Mass and other liturgical services. The colors are used to signify the nature of the liturgical season or feast being celebrated.

See [GIRM §346](https://www.liturgyoffice.org.uk/Resources/GIRM/Documents/GIRM.pdf#page=63).

## Liturgical Day

> A liturgical day denotes a specific and distinct day within the Church’s liturgical calendar, characterized by particular celebrations and observances. Each day is assigned a [rank](#rank) to indicate its significance, a specific [liturgical color](#liturgical-colors), and various properties that define its nature and all associated celebrations. The date of each liturgical day is determined according to the [Proper of Time](#proper-of-time), the [Proper of Saints](#proper-of-saints), and the rules governing the precedence of other liturgical days.

Reviewer: The link to Rank is now broken.

I have actually created a script to check for broken links in a folder. It is not tested much, should be definitely ported to TS, however, it works for `glossary.md`. IMO we should add it either to the `data:checks` script or create a custom one (e.g. `docs:checks`, but it should be run at least in GitHub Actions/CI. Here’s a first version of such script tested on the current version of the glossary.

The script found the following broken links:

```c#
glossary.md:38:176 - #rank
glossary.md:38:231 - #liturgical-colors
```

Required dependencies of `check-broken-links.js`:

```json
{
  "@jsdevtools/rehype-url-inspector": "2.0.2",
  "rehype-slug": "6.0.0",
  "rehype-stringify": "10.0.0",
  "remark-gfm": "4.0.0",
  "remark-parse": "11.0.0",
  "remark-rehype": "11.1.0",
  "unified": "11.0.5"
}
```

<details>
  <summary><code>check-broken-links.js</code></summary>

```js
import {access, readdir, readFile, stat} from 'node:fs/promises'
import {dirname, relative, resolve} from 'node:path'

import inspectUrls from '@jsdevtools/rehype-url-inspector'
import rehypeSlug from 'rehype-slug'
import rehypeStringify from 'rehype-stringify'
import remarkGfm from 'remark-gfm'
import remarkParse from 'remark-parse'
import remarkRehype from 'remark-rehype'
import {unified} from 'unified'

/** @param path - Folder or file with absolute or relative path */
const checkLinks = async (path) => {
  const files = await isDir(path) ? await getMarkdownFiles(path) : [path]

  for (const file of files) {
    await processFile(file, path)
  }
}

const getMarkdownFiles = async (directory) => {
  const files = await readdir(directory, {withFileTypes: true})
  const markdownFiles = []

  for (const file of files) {
    const fullPath = resolve(directory, file.name)

    if (file.isDirectory()) {
      const nestedFiles = await getMarkdownFiles(fullPath)
      markdownFiles.push(...nestedFiles)
    } else if (file.isFile() && file.name.endsWith('.md')) {
      markdownFiles.push(fullPath)
    }
  }

  return markdownFiles
}

const processFile = async (filePath, directory) => {
  const content = await readFile(filePath, 'utf8')
  const basePath = dirname(filePath)

  await unified()
    .use(remarkGfm)
    .use(remarkParse)
    .use(remarkRehype)
    .use(rehypeSlug)
    .use(inspectUrls, {
      inspectEach({node, root, url}) {
        // Note: We ignore external URLs (HTTP/S).
        if (url.match(/^(?!https?:\/\/).+/)) {
          // Check internal anchor links within the same file
          if (url.startsWith('#')) {
            if (!root.children.some((i) => i.properties?.id === url.substring(1))) {
              logBrokenLink({
                directory,
                filePath,
                position: node.position.start,
                url
              })
            }
          } else {
            // Check relative links (with or without anchor)
            const [relativePath, anchor] = url.split('#')

            // Check paths with and without `.md` extension
            // Note: Markdown files are usually referenced without the extension, other files require the extension specification.
            const targetPaths = [
              resolve(basePath, relativePath),
              resolve(basePath, `${relativePath}.md`)
            ]

            Promise.all(
              targetPaths.map(async (targetPath) => {
                try {
                  const pathStats = await stat(targetPath)
                  if (pathStats.isFile()) {
                    return targetPath
                  } else {
                    // Path is a directory
                    return undefined
                  }
                } catch {
                  // Path does not exist
                  return undefined
                }
              })
            )
              .then((resolvedPaths) => {
                const existingPath = resolvedPaths.find((p) => !!p)

                if (existingPath && anchor) {
                  checkAnchorInFile({
                    anchor,
                    directory,
                    filePath: existingPath,
                    originalUrl: url,
                    position: node.position.start,
                    sourceFile: filePath
                  })
                } else if (!existingPath) {
                  logBrokenLink({
                    directory,
                    filePath,
                    position: node.position.start,
                    url
                  })
                }
              })
              .catch(() => {
                logBrokenLink({
                  directory,
                  filePath,
                  position: node.position.start,
                  url
                })
              })
          }
        }
      }
    })
    .use(rehypeStringify)
    .process(content)
}

const checkAnchorInFile = async ({
  anchor,
  directory,
  filePath,
  originalUrl,
  position,
  sourceFile
}) => {
  const content = await readFile(filePath, 'utf8')
  const u = await unified()
    .use(remarkGfm)
    .use(remarkParse)
    .use(remarkRehype)
    .use(rehypeSlug)
    .use(rehypeStringify)
  const result = await u.run(u.parse(content))

  const anchorExists = result.children?.some((i) => i.properties?.id === anchor)

  if (!anchorExists) {
    logBrokenLink({
      directory,
      filePath: sourceFile,
      position,
      url: originalUrl
    })
  }
}

const logBrokenLink = ({directory, filePath, position, url}) => {
  const {line, column} = position
  console.log(`${directory}/${relative(directory, filePath)}:${line}:${column} - ${url}`)
}

const isDir = async (inputPath) => {
  try {
    const stats = await stat(inputPath)
    return stats.isDirectory()
  } catch {
    return false
  }
}

checkLinks(process.argv[2] || '.').catch((e) => {console.error(e)})
```

</details>

Reviewer 2: added in 36a39ac

A liturgical day generally aligns with a day in the Gregorian calendar but has some nuances: it typically begins at First Vespers or Vigils (if observed), or otherwise at midnight, and ends at the beginning of the following liturgical day. If a liturgical day begins before midnight, it remains identified as a single liturgical day, even if it spans two civil days. A liturgical day is identified by a Gregorian date (considering the majority of the liturgical day that occurs within a given civil day).

> In Romcal, it is this concept of the liturgical day that is calculated within a liturgical calendar.

Reviewer: ```suggestion
In romcal, it is this concept of the liturgical day that is calculated within a liturgical calendar.
```

> [!NOTE]
> A liturgical day should not be confused with a celebration (see [celebration](#celebration)).

## Liturgical Period

> In Romcal, in addition to the liturgical seasons, liturgical periods provide further detail and metadata about the current liturgical period. For example, the octaves of Christmas and Easter, the period before and after Epiphany, Holy Week, etc. They also denote periods traditionally observed in the Catholic Church and by religious communities (especially for Marian hymns), including the period from the Baptism of Christ to the Presentation of Jesus in the Temple, and from the Presentation of Jesus in the Temple to Holy Thursday.

Reviewer: ```suggestion
In romcal, in addition to the liturgical seasons, liturgical periods provide further detail and metadata about the current liturgical period. For example, the octaves of Christmas and Easter, the period before and after Epiphany, Holy Week, etc. They also denote periods traditionally observed in the Catholic Church and by religious communities (especially for Marian hymns), including the period from the Baptism of Christ to the Presentation of Jesus in the Temple, and from the Presentation of Jesus in the Temple to Holy Thursday.
```

> 

Reviewer: If we update the glossary item hedding for the _General Instruction of the Liturgy of the Hours_, we should update the link to it to `#general-instruction-of-the-liturgy-of-the-hours`.

```suggestion

## Liturgy of the Hours

The Liturgy of the Hours (LotH), or the Divine Office, is a set of prayers prescribed by the Catholic Church to be recited at specific times of the day by clergy, religious, and laity. It consists of psalms, hymns, readings, and other prayers, structured to sanctify the day and mark the passage of time. The principal hours are Lauds (Morning Prayer), Vespers (Evening Prayer), and Compline (Night Prayer), with additional hours such as Matins (Office of Readings), Terce (Mid-Morning Prayer), Sext (Midday Prayer), and None (Mid-Afternoon Prayer). The LotH aims to fulfil St Paul’s exhortation to ‘pray without ceasing’ (1 Thessalonians 5:17).

See also:

- [General Instruction of the Liturgy of the Hours](#general-instruction-of-the-liturgy-of-the-hours-gilh).

```

See [GNLYC §59](https://www.liturgyoffice.org.uk/Resources/GIRM/Documents/GNLY.pdf#page=14).

## Proper of Saints

The Proper of Saints (or Sanctoral) consists of the fixed feasts, celebrated on the very same date each year (no matter what the day of the week),including Christmas and all the saints’ days.

## Proper of Time

The Proper of Time (or Temporal) consists of the movable feasts, most of them keyed to Easter (which falls on a different Sunday every year), including Ascension, Pentecost, and so on.

> [!NOTE]
> We avoid using the term Temporal to prevent potential confusion with the `Temporal` object in ECMAScript, which is still in a [proposal stage](https://tc39.es/proposal-temporal/docs/).

## Psalter Week Cycle

The four-week cycle of the psalter is coordinated with the liturgical year in such a way that on the First Sunday of Advent, the First Sunday in Ordinary Time, the First Sunday of Lent, and Easter Sunday the cycle is always begun again with Week 1 (others being omitted when necessary).

See [GILH §133](https://www.liturgyoffice.org.uk/Resources/Rites/GILH.pdf#page=26).

> ## Liturgical Rank

Reviewer: @emagnier, if you want to keep the `Liturgical` word in the title, the glossary item should be moved to keep the alphabetical order.

Anyway, we should sort the glossary programmatically, probably as part of the `data:checks` script, or at least in GitHub Actions/CI. Here’s a first version of such script tested on the current version of the glossary.

Required dependencies of `sort-glossary.js`:

```json
{
  "remark": "15.0.1",
  "remark-gfm": "4.0.0",
  "remark-parse": "11.0.0",
  "remark-stringify": "11.0.0"
}
```

<details>
  <summary><code>sort-glossary.js</code></summary>

```js
import {readFile, writeFile} from 'node:fs'
import {resolve} from 'node:path'

import {remark} from 'remark'
import gfm from 'remark-gfm'
import parse from 'remark-parse'
import stringify from 'remark-stringify'

// TODO: Change the path of `fIn` to `glossary.md`.
// TODO: Consider removing `fOut` (we should replace the original file) and renaming `fIn`t to `file` or `glossary`.
const fIn = resolve('./romcal-glossary-sort-test.md')
const fOut = resolve('./romcal-glossary-sort-test-out.md')

/** Sort second-level headings alphabetically */
const sortHeadings = tree => {
  const sortedHeadings = []
  let currentHeading = null

  tree.children.forEach((node) => {
    if (node.type === 'heading' && node.depth === 2) {
      if (currentHeading) {
        sortedHeadings.push(currentHeading)
      }
      currentHeading = {
        heading: node,
        content: []
      }
    } else if (currentHeading) {
      currentHeading.content.push(node)
    }
  })

  if (currentHeading) {
    sortedHeadings.push(currentHeading)
  }

  sortedHeadings.sort((a, b) => {
    const headingA = a.heading.children.map((child) => child.value).join('')
    const headingB = b.heading.children.map((child) => child.value).join('')
    return headingA.localeCompare(headingB)
  })

  const newChildren = []
  sortedHeadings.forEach((section) => {
    newChildren.push(section.heading)
    section.content.forEach((node) => {
      newChildren.push(node)
    })
  })

  tree.children = newChildren
}

/** Read, sort and write the Markdown file */
const processFile = (fIn, fOut) => {
  readFile(fIn, 'utf8', (e, data) => {
    if (e) {
      console.error(e)
      return
    }

    remark()
      .use(parse)
      .use(gfm)
      .use(() => tree => sortHeadings(tree))
      .use(stringify, { bullet: '-', fence: '`', fences: true, incrementListMarker: false, listItemIndent: 'one' })
      .process(data, (e, file) => {
        if (e) {
          console.error(e)
          return
        }

        // Note: We need to remove the escapes from alerts, as `remark-gfm` does not support these yet, see https://github.com/remarkjs/remark-gfm/issues/66.
        writeFile(fOut, String(file).replace(/^> \\\[!(NOTE|TIP|IMPORTANT|WARNING|CAUTION)\][ \t]*$/gm, '> [!$1]'), 'utf8', e => {
          if (e) {
            console.error(e)
            return
          }

          console.log('Glossary items are sorted successfully.')
        })
      })
  })
}

processFile(fIn, fOut)
```

</details

Author: With some perspective now, I prefer that we use short names and not necessarily prefix with the word `Liturgy`. Since this project is already related to liturgy, it would be redundant. And it's also better to have short names in the codebase.

Unless it could cause confusion (for example, it's better to keep `LiturgicalDay`, because `Day` alone indicates a civil day).

So here, I would prefer to revert to `Rank` rather than `LiturgicalRank`.

Reviewer 2: working through this script you implemented @tukusejssirs (which has been added on this PR). it doesn't handle h1's, so the top-level header of "Glossary" is removed upon sort.... just letting you know, i'm working on a fix.

Reviewer 2: i've got it sorted now, i've also pushed the code for warning you that it's not sorted when you try to commit

# Glossary

This glossary provides definitions of terms used in the context of this romcal project. The terms are listed in alphabetical order.

The structure of a glossary items is as follows:

- the term name in a heading of the second level, without any abbreviation:
  - abbreviations, acronyms, initialisms, etc., should be made a separate item with `See [Full Term Name](#full-term-name).` (change the link name and anchor as required);
  - the headings should be ordered alphabetically;
- definition of the term:
  - ideally, it should full in a single, short paragraph;
  - provide an abbreviation (if it is used) after the first mention of the term;
- add a list named `See also:` where the synonyms and related items would be provided (provide as many links as required; prefer links to the glossary items).

## Abbreviations

### GILH

> See #general-instruction-of-the-liturgy-of-the-hours.

Reviewer: ```suggestion
See [General Instruction of the Liturgy of the Hours](#general-instruction-of-the-liturgy-of-the-hours).
```

### GIRM

> See #general-instruction-on-the-roman-missal.

Reviewer: ```suggestion
See [General Instruction on the Roman Missal](#general-instruction-on-the-roman-missal).
```

### GNLYC

> See #general-norms-for-the-liturgical-year-and-the-calendar.

Reviewer: ```suggestion
See [General Norms for the Liturgical Year and the Calendar](#general-norms-for-the-liturgical-year-and-the-calendar).
```

See [GILH §133](https://www.liturgyoffice.org.uk/Resources/Rites/GILH.pdf#page=26).

## Rank

The rank of a liturgical celebration is a measure of its importance in the liturgical calendar. The rank of a celebration determines how it should be observed when it falls on the same day as another celebration.

There are five categories of ranks: solemnities, Sunday, feasts, memorials, weekday. A memorial can be obligatory or optional.

See [GNLYC §3-16](https://www.liturgyoffice.org.uk/Resources/GIRM/Documents/GNLY.pdf#page=5).

## Sanctoral

See [Proper of Saints](#proper-of-saints).

## Season

Liturgical seasons are the periods of time in the liturgical year that are marked by a particular liturgical color and have a particular focus. The liturgical seasons are Advent, Christmastide, Lent, Easter Time, and Ordinary Time. The Ordinary Time is divided into two parts: the first part begins on the Monday after the Baptism of the Lord and ends on the day before Ash Wednesday; the second part begins on the Monday after Pentecost and ends on the Saturday before the First Sunday of Advent.

## Sunday Cycle

> A three-year cycle for Sunday Mass readings (and some solemnities), designated by A, B, or C. Each cycle begins on the First Sunday of Advent of the previous civil year and ends on Saturday after the Christ the King Solemnity. The cycles follow each other in alphabetical order. C year is always divisible by 3, A has remainder of 1, and B remainder of 2.

Reviewer: I believe that `Year C`/`Cycle I` (when specifying the letter/number is considered a title/name of that cycle, thus it should be capitalised.

However, @TobiTenno, correct me if I am wrong. :pray:

```suggestion
A three-year cycle for Sunday Mass readings (and some solemnities), designated by A, B, or C. Each cycle begins on the First Sunday of Advent of the previous civil year and ends on Saturday after the Christ the King Solemnity. The cycles follow each other in alphabetical order. Year C is always divisible by 3, A has remainder of 1, and B remainder of 2.
```

Reviewer 2: it's talking about the numerical property "year" of C, because `Year C's numerical value` would be a bit more wordy than is worth it to me. also, it doesn't really matter to me in this case

Reviewer: If `year` refers to a property name, IMO it should be styled as a code (using backticks), however, in this case it specifically talks about Year C (or _every Year C year_ :wink:).

Now, I think that sentence should be rephrased. How about this?

```suggestion
A three-year cycle for Sunday Mass readings (and some solemnities), designated by A, B, or C. Each cycle begins on the First Sunday of Advent of the previous civil year and ends on Saturday after the Christ the King Solemnity. The cycles follow each other in alphabetical order. Year C is always the year divisible by 3, while Year A leaves a remainder of 1, and Year B a remainder of 2.
```

Or even add `Gregorian`:

```suggestion
A three-year cycle for Sunday Mass readings (and some solemnities), designated by A, B, or C. Each cycle begins on the First Sunday of Advent of the previous civil year and ends on Saturday after the Christ the King Solemnity. The cycles follow each other in alphabetical order. Year C is always the Gregorian year divisible by 3, while Year A leaves a remainder of 1, and Year B a remainder of 2.
```

Reviewer 2: i think we can discuss these semantics later and all this falls into "further iterations" and doesn't need to be exact this second.

Reviewer: Okay.

# Glossary

> Definition of commonly used terms in the context of this Romcal project. The terms are listed in alphabetical order.

Reviewer: ```suggestion
This glossary provides definitions of terms used in the context of this romcal project. The terms are listed in alphabetical order.
```

> Definition of commonly used terms in the context of this Romcal project. The terms are listed in alphabetical order.

Reviewer: I think we should append to each glossary item (term) a list of synonyms and a list of related items. OFC, these lists should be added only if there is at least one list item, otherwise the list should be removed.

Also specify a abbreviation/acronym/initialism when need ed after the first usage of the term in the term’s definition. Never specify it in a glosaary item name.

Moreover, I think that each synonym (incl abbreviations) should be made a glossary item with only a link to the main item (e.g. `See [Main Item](#main-item).`). If a term is ambiguous, we could provide multiple links. We already do this, albeit only seldom.For example:

- `Rank` → `Liturgical Rank`;
- `Color` → `Liturgical Color`;
- `Calendar` → `Liturgical Calendar`;
- `GIRM` → `General Instruction on the Roman Missal`;
- `Breviary` → `Liturgy of the Hours`;
- `Divine Office` → `Liturgy of the Hours`.

We also should use [alerts](https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax#alerts).

Terms to add:

- `Occurrence`;
- `Psalter`;
- `Liturgy of the Hours`.

I could add some actual suggestions for these, hpwever, first I’d like to hear some confirmation that you agree with me. @emagnier and @TobiTenno.

All rules we agree upon should be added under the `Glossary` heading in this file.

Author: For the acronyms/abbreviation, I think it's important to keep them in the title to keep them prominent (especially GIRM, GNLYC, GILH). Since these acronyms are what we'll use elsewhere to refer to this document, readers of the documentation need to quickly associate the definition with the acronym. What do you think?

For adding the word `Liturgical` in front of several other words, why not? If it might cause confusion if `Liturgical` isn't present. Ideally, the term used should match the corresponding `type`/`class`/`enum` in the codebase. So we need something explicit, but not too long.

For the rest, I'm aligned with you.

Author: By the way, I applied your suggestion to change `Rank` to `Liturgical Rank`. However, upon reflection, I find it a bit verbose.

Reviewer: > For the acronyms/abbreviation, I think it's important to keep them in the title to keep them prominent (especially GIRM, GNLYC, GILH). Since these acronyms are what we'll use elsewhere to refer to this document, readers of the documentation need to quickly associate the definition with the acronym. What do you think?

While I agree that (at least) those three initialism are quite important to have them in the glossary, however, I disagree to put them to the headings of the respective glossary items. I propose to add a separate glossary items for these (see also [this comment of mine](https://github.com/romcal/romcal/pull/586#discussion_r1659886469)).

> For adding the word `Liturgical` in front of several other words, why not? If it might cause confusion if `Liturgical` isn't present.

Unlike for the problem discussed above, I don’t consider removing (not adding) the `liturgical` word to any of the mentioned words a big issue. Still, I think it would be better to add it to some/most of the mentioned words because the words have multiple meanings or usages in different context, however, I get it that romcal defines the liturgical context.

Therefore, add `Liturgical` where you agree with me, remove it where you disagree with me. :wink:

> Ideally, the term used should match the corresponding `type`/`class`/`enum` in the codebase.

I think it should be the opposite way: the names of code types/classes/enums/variables should match the name used by the Church in (American) English nowadays. :wink: Where there are discrepancies, we should update the code, not the terms. :smiley:

> So we need something explicit, but not too long.

IMHO, the glossary item names should be short, but sufficiently specific.

> For the rest, I'm aligned with you.

Thanks! :pray:

I’ll re-review the glossary and add some more suggestions. :wink:

> ## Celebration

Reviewer: ```suggestion
## Liturgical Celebration
```

Author: Actually, I'm questioning the relevance of prefixing with the word "Liturgical" for "Celebration". Since in the context of this project, a celebration is inherently liturgical. And when we eventually have objects of type `Celebration` in the codebase, it's also good to keep names from being too long.

Reviewer 2: I agree, I read it a few times, thought about it.

Reducing verbosity spread on names becomes a real problem when you've got as many types as we do

Reviewer: > Reducing verbosity spread on names becomes a real problem when you've got as many types as we do

I personally don’t think the verbosity and length are an issue for this glossary of ours. It should contain as many terms as we / the Church use (-s). If we (can) use multiple terms for the same thing, we should add a glossary item for each, all (but one) linking to the main one (e.g. `Feria` should refer to `Weekday` or `GIRM` to `General Instruction on the Roman Missal`).

Reviewer 2: Then why not link "Liturgical Celebration" to "Celebration" instead of extending the main definition, requiring a reordering as well

Reviewer: > Then why not link "Liturgical Celebration" to "Celebration" instead of extending the main definition

That would be another option. However, it is okay as it is, no need to accept my suggestion to prepend `Liturgical` to many terms.

> requiring a reordering as well

I don’t see any downside in this, even more when we have now a script to do the reordering automatically. :wink:

A celebration is a liturgical or sacramental event, typically consisting of rituals, readings, prayers, and often the Eucharist (Mass), whether or not it is associated with a liturgical day.

> Note: Currently, Romcal does not provide specific information about celebrations, such as the list of Masses for a given day or the hours of the Divine Office. Instead, it indicates unique liturgical days determined within a liturgical calendar. However, the ability to list these celebrations—linked to a liturgical day or corresponding to a civil day (such as Vigil Masses)—is a feature that may be added in the future.

Reviewer: ```suggestion
> [!NOTE]
> Currently, Romcal does not provide specific information about celebrations, such as the list of Masses for a given day or the hours of the Divine Office. Instead, it indicates unique liturgical days determined within a liturgical calendar. However, the ability to list these celebrations—linked to a liturgical day or corresponding to a civil day (such as Vigil Masses)—is a feature that may be added in the future.
```

> ## General Instruction of the Liturgy of the Hours (GILH)

Reviewer: ```suggestion
## General Instruction of the Liturgy of the Hours
```

Author: Here, I think it's important to keep the acronym in the title to keep it prominent. Because this acronym is what we'll use elsewhere to refer to this document.

Reviewer 2: agreed

Reviewer: Well, I consider a glossary item to be similar to a dictionary entry from the formal/structure point of view: the headword in a dictionary entry should not include anything else but the word defined only.

Same here: while we don’t place definition of a glossary item on the same line as the term name, the term name (heading in our case) should not include anything but the term name.

For abbreviations/acronyms/initialisms we should add an additional glossary item where we point the reader to the proper place using `See [Some Term](#some-term)`. Etienne, you already do this for some of the items, e.g. [Sanctoral](https://github.com/romcal/romcal/blob/e17e05a96a63e98771e8a6c0d114f92cb59b0297/docs/glossary.md#sanctoral) refers to the [Proper of Saints](https://github.com/romcal/romcal/blob/e17e05a96a63e98771e8a6c0d114f92cb59b0297/docs/glossary.md#proper-of-saints).

Therefore, I don’t think it is a good idea to append abbreviations/acronyms/initialisms to main glossary item names. However, we should add it after the first use of the term in its definition (again, like you’ve already done it in [Proper of Saints](https://github.com/romcal/romcal/blob/e17e05a96a63e98771e8a6c0d114f92cb59b0297/docs/glossary.md#proper-of-saints)).

Reviewer 2: i'll go through (in a few hours) and add an upper section of abbreviations to cover our concerns about abbreviations being lost, that way they don't have to be in these titles and can be linked as the abbreviations in the titles with a `see -> #longer-title-form` in the glossary.

The General Instruction of the Liturgy of the Hours (GILH) is the document that provides the general instructions for the Liturgy of the Hours. It is an official liturgical document that was promulgated by the Congregation for Divine Worship and the Discipline of the Sacraments.

> An English translation of the GILH can be found [here](https://divineoffice.org/general-instructions/).

Reviewer: I prefer a [PDF form in a better visual format](https://www.liturgyoffice.org.uk/Resources/Rites/GILH.pdf), however, it seems to be anglicised (e.g. it uses -our suffices, but it uses -ize suffices).

```suggestion
An English translation of the GILH can be found [here](https://www.liturgyoffice.org.uk/Resources/Rites/GILH.pdf).
```

> ## General Instruction on the Roman Missal (GIRM)

Reviewer: ```suggestion
## General Instruction on the Roman Missal
```

Author: Here, I think it's important to keep the acronym in the title to keep it prominent. Because this acronym is what we'll use elsewhere to refer to this document.

Reviewer 2: i'm not gonna comment on all of them, but i'm gonna resolve the ones that are all this same change.

The General Instruction on the Roman Missal (GIRM) is the detailed document that provides the rubrics for the celebration of the Mass. It is an official liturgical document that was promulgated by the Congregation for Divine Worship and the Discipline of the Sacraments.

> An English translation of the GIRM can be found [here](https://www.catholicculture.org/culture/library/view.cfm?recnum=337).

Reviewer: The link contains the GIRM from the Roman Missal, ed typ altera. We should provide a link to one from ed typ tertia.

We could link to one provided by the [Holy See](https://www.vatican.va/roman_curia/congregations/ccdds/documents/rc_con_ccdds_doc_20030317_ordinamento-messale_en.html), however, I think the best one would be would have links or CSS IDs for each paragraph of the GIRM. As I could not find such, I prefer linking to a PDF (like [this one](https://www.liturgyoffice.org.uk/Resources/GIRM/Documents/GIRM.pdf)) and use `page=$pg_num` when provoding links to a part of the GIRm.

```suggestion
An English translation of the GIRM can be found [here](https://www.liturgyoffice.org.uk/Resources/GIRM/Documents/GIRM.pdf).
```

> ## General Norms for the Liturgical Year and the Calendar (GNLYC)

Reviewer: ```suggestion
## General Norms for the Liturgical Year and the Calendar 
```

Author: Here, I think it's important to keep the acronym in the title to keep it prominent. Because this acronym is what we'll use elsewhere to refer to this document.

Reviewer 2: agreed

The General Norms for the Liturgical Year and the Calendar (GNLYC) is the document that provides the general norms for the liturgical year and the calendar. It is an official liturgical document that was promulgated by the Congregation for Divine Worship and the Discipline of the Sacraments.

> An English translation of the GNLYC can be found [here](https://www.catholicculture.org/culture/library/view.cfm?id=10842).

Reviewer: I prefer a [PDF form in a better visual format](https://www.liturgyoffice.org.uk/Resources/GIRM/Documents/GNLY.pdf), however, it seems to be anglicised (e.g. it uses -ise suffices).

```suggestion
An English translation of the GNLYC can be found [here](https://www.liturgyoffice.org.uk/Resources/GIRM/Documents/GNLY.pdf).
```

> ## Liturgical Colors

Reviewer: ```suggestion
## Liturgical Color
```

The liturgical colors are the colors used in the liturgical vestments and altar cloths for the celebration of the Mass and other liturgical services. The colors are used to signify the nature of the liturgical season or feast being celebrated.

> See GIRM §346.

Reviewer: ```suggestion
See [GIRM §346](https://www.liturgyoffice.org.uk/Resources/GIRM/Documents/GIRM.pdf#page=63).
```

## Liturgical Day

> A liturgical day denotes a specific and distinct day within the Church's liturgical calendar, characterized by particular celebrations and observances. Each day is assigned a rank to indicate its significance, a specific liturgical color, and various properties that define its nature and all associated celebrations. The date of each liturgical day is determined according to the Proper of Time, the Proper of Saints, and the rules governing the precedence of other liturgical days.

Reviewer: Note that if we update the `Rank` item name to `Liturgical Rank`, it should be reflected in the suggestion below.

```suggestion
A liturgical day denotes a specific and distinct day within the Church’s liturgical calendar, characterized by particular celebrations and observances. Each day is assigned a [rank](#rank) to indicate its significance, a specific [liturgical color](#liturgical-colors), and various properties that define its nature and all associated celebrations. The date of each liturgical day is determined according to the [Proper of Time](#proper-of-time), the [Proper of Saints](#proper-of-saints), and the rules governing the precedence of other liturgical days.
```

A liturgical day generally aligns with a day in the Gregorian calendar but has some nuances: it typically begins at First Vespers or Vigils (if observed), or otherwise at midnight, and ends at the beginning of the following liturgical day. If a liturgical day begins before midnight, it remains identified as a single liturgical day, even if it spans two civil days. A liturgical day is identified by a Gregorian date (considering the majority of the liturgical day that occurs within a given civil day).

In Romcal, it is this concept of the liturgical day that is calculated within a liturgical calendar.

> Note: A liturgical day should not be confused with a celebration (see celebration).

Reviewer: ```suggestion
> [!NOTE]
> A liturgical day should not be confused with a celebration (see [celebration](#celebration)).
```

> ## Period

Reviewer: I suggest to use `Liturgical` word in the title.

```suggestion
## Liturgical Period
```

In Romcal, in addition to the liturgical seasons, liturgical periods provide further detail and metadata about the current liturgical period. For example, the octaves of Christmas and Easter, the period before and after Epiphany, Holy Week, etc. They also denote periods traditionally observed in the Catholic Church and by religious communities (especially for Marian hymns), including the period from the Baptism of Christ to the Presentation of Jesus in the Temple, and from the Presentation of Jesus in the Temple to Holy Thursday.

## Precedence

The principle of precedence is used to determine which liturgical celebration should be observed when two or more celebrations fall on the same day. The principle of precedence is based on the liturgical rank of the celebrations.

> See GNLYC §59.

Reviewer: ```suggestion
See [GNLYC §59](https://www.liturgyoffice.org.uk/Resources/GIRM/Documents/GNLY.pdf#page=14).
```

> ## Proper of Saints (or Sanctoral)

Reviewer: ```suggestion
## Proper of Saints
```

> The "Proper of Saints" (or "Sanctoral") consists of the fixed feasts, celebrated on the very same date each year (no matter what the day of the week),including Christmas and all the saints' days.

Reviewer: ```suggestion
The Proper of Saints (or Sanctoral) consists of the fixed feasts, celebrated on the very same date each year (no matter what the day of the week),including Christmas and all the saints’ days.
```

> ## Proper of Time (or Temporal)

Reviewer: ```suggestion
## Proper of Time
```

> The "Proper of Time" (or "Temporal") consists of the movable feasts, most of them keyed to Easter (which falls on a different Sunday every year), including Ascension, Pentecost, and so on.

Reviewer: ```suggestion
The Proper of Time (or Temporal) consists of the movable feasts, most of them keyed to Easter (which falls on a different Sunday every year), including Ascension, Pentecost, and so on.
```

> Note: we avoid using the term "Temporal" to prevent potential confusion with the `Temporal` object in JavaScript, which is still in a [proposal stage](https://tc39.es/proposal-temporal/docs/).

Reviewer: I also think we should replace `JavaScript` with `ECMAScript`.

```suggestion
> [!NOTE]
> We avoid using the term Temporal to prevent potential confusion with the `Temporal` object in JavaScript, which is still in a [proposal stage](https://tc39.es/proposal-temporal/docs/).
```

Author: ```suggestion
> [!NOTE]
> We avoid using the term Temporal to prevent potential confusion with the `Temporal` object in ECMAScript, which is still in a [proposal stage](https://tc39.es/proposal-temporal/docs/).
```

## Psalter Week Cycle

The four-week cycle of the psalter is coordinated with the liturgical year in such a way that on the First Sunday of Advent, the First Sunday in Ordinary Time, the First Sunday of Lent, and Easter Sunday the cycle is always begun again with Week 1 (others being omitted when necessary).

> See GILH §133.

Reviewer: ```suggestion
See [GILH §133](https://www.liturgyoffice.org.uk/Resources/Rites/GILH.pdf#page=26).
```

> ## Rank

Reviewer: I suggest to use `Liturgical` word in the title.

```suggestion
## Liturgical Rank
```

The rank of a liturgical celebration is a measure of its importance in the liturgical calendar. The rank of a celebration determines how it should be observed when it falls on the same day as another celebration.

There are five categories of ranks: solemnities, Sunday, feasts, memorials, weekday. A memorial can be obligatory or optional.

> See GNLYC §3-16.

Reviewer: ```suggestion
See [GNLYC §3-16](https://www.liturgyoffice.org.uk/Resources/GIRM/Documents/GNLY.pdf#page=5).
```

## Sanctoral

> See [Proper of Saints]().

Reviewer: ```suggestion
See [Proper of Saints](#proper-of-saints).
```

## Season

Liturgical seasons are the periods of time in the liturgical year that are marked by a particular liturgical color and have a particular focus. The liturgical seasons are Advent, Christmastide, Lent, Easter Time, and Ordinary Time. The Ordinary Time is divided into two parts: the first part begins on the Monday after the Baptism of the Lord and ends on the day before Ash Wednesday; the second part begins on the Monday after Pentecost and ends on the Saturday before the First Sunday of Advent.

> ## Sunday Cycles

Reviewer: I would use the term on singular, unless it has a unique/special meaning in plural.

```suggestion
## Sunday Cycle
```

A three-year cycle for Sunday Mass readings (and some solemnities), designated by A, B, or C. Each cycle begins on the First Sunday of Advent of the previous civil year and ends on Saturday after the Christ the King Solemnity. The cycles follow each other in alphabetical order. C year is always divisible by 3, A has remainder of 1, and B remainder of 2.

## Temporal

> See [Proper of Time]().

Reviewer: ```suggestion
See [Proper of Time](#proper-of-time).
```

> ## Weekday Cycles

Reviewer: I would use the term on singular, unless it has a unique/special meaning in plural.

```suggestion
## Weekday Cycle
```
